build-tools/packages/build-infrastructure/src/releaseGroup.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { type ReleaseGroupDefinition, matchesReleaseGroupDefinition } from "./config.js";
+import type {
+	IPackage,
+	IReleaseGroup,
+	IWorkspace,
+	PackageName,
+	ReleaseGroupName,
+} from "./types.js";
+
+export class ReleaseGroup implements IReleaseGroup {
+	public readonly name: ReleaseGroupName;
+	public readonly adoPipelineUrl: string | undefined;
+	public constructor(
+		name: string,
+		releaseGroupDefinition: ReleaseGroupDefinition,
+		public workspace: IWorkspace,
+		public readonly rootPackage?: IPackage,
+	) {
+		this.name = name as ReleaseGroupName;
+		this.adoPipelineUrl = releaseGroupDefinition.adoPipelineUrl;
+		this.packages = workspace.packages
+			.filter((pkg) => matchesReleaseGroupDefinition(pkg, releaseGroupDefinition))
+			.map((pkg) => {
+				// update the release group in the package object so we have an easy way to get from packages to release groups
+				pkg.releaseGroup = this.name;
```

Comment:
Yes, but I can't figure out how to do it at Package creation time. Packages really "belong" to a workspace, so when workspaces are instantiated, the packages have to be instantiated. Then release groups are defined/instantiated after workspaces. So when the packages are instantiated, there is no instantiated release group.

I'll keep thinking about this. The problem I am trying to avoid is having properties that can be undefined, which requires that they be set in the constructor. Maybe I could use getters + null checks in the class implementation? Open to other ideas.

Reply:
The current "workaround" are these two functions on IFluidRepo:

```ts
	/**
	 * Returns the {@link IReleaseGroup} associated with a package.
	 */
	getPackageReleaseGroup(pkg: Readonly<P>): Readonly<IReleaseGroup>;

	/**
	 * Returns the {@link IWorkspace} associated with a package.
	 */
	getPackageWorkspace(pkg: Readonly<P>): Readonly<IWorkspace>;
```

Not sure that second one is really needed because IPackage already has a `workspace` property that is the same thing.

Reply:
I removed that second function - it was indeed superfluous.